cores/esp8266/core_esp8266_postmortem.cpp

```diff
@@ -322,5 +403,199 @@ void __stack_chk_fail(void) {
+asm(  // 32 bytes MAX, using 31
+    ".section     .text.postmortem_kernel_exception_handler,\"ax\",@progbits\n\t"
+    ".align       4\n\t"
+    ".type        postmortem_kernel_exception_handler, @function\n\t"
+    "\n"
+"postmortem_kernel_exception_handler:\n\t" // Copy destination _KernelExceptionVector
+    "wsr.excsave1 a0\n\t"
+
+"postmortem_kernel_exception_handler_p3:\n\t"
+    "addi         a0,     a1,     -32\n\t"
+    "s32i         a1,     a0,     24\n\t"
+    "mov.n        a1,     a0\n\t"
+    "xsr.excsave2 a0\n\t"                 // stash pointer to stack data in excsave2
+    "s32i         a0,     a1,     12\n\t" // and save previous
+    "rsr.excsave1 a0\n\t"
+    "s32i         a0,     a1,     28\n\t"
+    "rsr.epc1     a0\n\t"
+    "s32i         a0,     a1,     20\n\t"
+    // save current exccause
+    "rsr.exccause a0\n\t"
+    "j            . - 54\n\t"   // continue at postmortem_debug_exception_vector_part2
+
+    ".size postmortem_kernel_exception_handler, .-postmortem_kernel_exception_handler\n\t"
+);
+
+void postmortem_double_exception_handler(void);
+asm(  // 16 byte MAX, using 14
+    ".section     .text.postmortem_double_exception_handler,\"ax\",@progbits\n\t"
+    ".align       4\n\t"
+    ".type        postmortem_double_exception_handler, @function\n\t"
+    "\n"
+"postmortem_double_exception_handler:\n\t" // Copy destination _DoubleExceptionVector
+    // Common code at _KernelExceptionVector, which is -64 bytes from
+    // _DoubleExceptionVector.
+    "j             .  - 64\n\t" // continue at _KernelExceptionVector
+
+    // finish here from _DebugExceptionVector part 2
+    // _DoubleExceptionVector part 2
+"postmortem_double_exception_handler_part2:\n\t"
+    // User mode, Clear EXCM, and block all interrupts
+    "movi.n       a0,    0x02F\n\t"
+    "wsr.ps       a0\n\t"
+    "rsync\n\t"
+    // Let _UserExceptionVector finish processing the exception as an
+    // exception(0). Adjust results at __wrap_system_restart_local so that `ESP
+    // Exception Decoder` reports the state before this `ill` instruction.
+    "ill\n\t"
+    ".size postmortem_double_exception_handler, .-postmortem_double_exception_handler\n\t"
+);
+
+/*
+  Of the Exception Cause table's 64 entries, most are set to
+  `_xtos_unhandled_exception`. And I assume most of these never occur; however,
+  at lesat one does exccause 20. This one occurs on a call to a NULL pointer. This
+  can happen when a function prototype is given the weak attribute but never
+  defined, or forgetting to test a call back function pointer for NULL before
+  calling, etc.
+
+  A problem with the `_xtos_unhandled_exception` function is that it handles the
+  event with a Debug Exception BP. If you don't have GDB running you see a HWDT
+  reset. If interrupts are at INTLEVEL 2 or higher (a BP instruction becomes a
+  NOP), you still see a HWDT reset regardless of running GDB.
+*/
+typedef void (* _xtos_handler)(void);
```

Comment:
Prefer using= over typedef for type definitions. In the case of a function ptr:
```cpp
using _xtos_handler = void (*)(void);
```

Reply:
okay

Reply:
I thought I had this covered, but I don't think so anymore. I think my problem stems from trying to reference xtensa sources that I can find on the Internet to fill in some of these missing function types. In the case of `_xtos_handler`, I am having trouble. They applied it to two very different function tables and it matches neither exactly. 
1) The first table is a set of ASM functions that do not comply with "C" conventions.
   * For `.cpp` files the typedef boils down to `typedef void (*_xtos_handler)(...);`
   * `extern _xtos_handler _xtos_exc_handler_table[];`
   * Of the set of ASM functions that are used in this table, there is a "C" wrapper function (`_xtos_c_wrapper_handler`) that references the 2nd table.
2) The 2nd table of "C" styled exception handling functions, defines its entries with the same typedef.
   * `extern _xtos_handler _xtos_c_handler_table[];`
   * for the functions in this table I think this would better match the function call:
     * `typedef void (_xtos_c_exception_handler_func)(struct __exception_frame *ef, int cause);`
     * `typedef _xtos_c_exception_handler_func *_xtos_c_handler;`
     * `extern _xtos_c_handler _xtos_c_handler_table[];`

My quandary is: 
 1) should I keep the typedef's I found and cast everything to fit? or
 2) Ignore the existing ones and hope those files never need to be added to this code base and define new ones?
     * for the first table I don't see a perfect fit for a non-"C" ASM. I think the current typedef would be fine.

Also affected: https://github.com/esp8266/Arduino/pull/7060#issue-370664751